project1-sps/script.js

```diff
@@ -0,0 +1,227 @@
+var roll3Number = function () {
+  // generate decimal 0 <= x < 3
+  var randomDecimal = Math.random() * 3;
+
+  // generate integer 0 <= x <= 2
+  var randomInteger = Math.floor(randomDecimal);
+
+  var chosenNumber = randomInteger + 1;
+  return chosenNumber;
+};
+
+let userName = "";
+let gameState = "start";
+let totalGames = 0;
+let totalWin = 0;
+let totalLose = 0;
+let totalDraw = 0;
+let currentMode = 0;
+
+var main = function (input) {
+  let message;
+
+  // ask for name
+  if (gameState == "start") {
+    gameState = "username";
+    message = "Hi! Welcome to Scissors-Paper-Stone. Please key in your name.";
+    return message;
+  }
+
+  // ask for mode
+  if (gameState == "username") {
+    userName = input;
+    gameState = "mode";
+    message =
+      `Nice to meet you, ${userName}. Please select which mode (1, 2, 3 or 4) will you be playing.` +
+      "<br>" +
+      "<br>" +
+      "1: Manual Normal mode" +
+      "<br>" +
+      "2: Manual Reverse mode" +
+      "<br>" +
+      "3: AI Normal mode" +
+      "<br>" +
+      "4: AI Reverse mode";
+    return message;
+  }
+
+  // state mode and start playing
+  if (gameState == "mode") {
+    if (input == 1) {
+      gameState = "play1";
+      currentMode = "Manual Normal Mode";
+      message =
+        `Nice to meet you, ${userName}. You are playing manual normal mode.` +
+        "<br>" +
+        `Please choose between scissors, paper and stone.`;
+      return message;
+    } else if (input == 2) {
+      gameState = "play2";
+      currentMode = "Manual Reverse Mode";
+      message =
+        `Nice to meet you, ${userName}. You are playing manual reverse mode.` +
+        "<br>" +
+        `Please choose between scissors, paper and stone.`;
+      return message;
+    } else if (input == 3) {
+      gameState = "playAI1";
+      currentMode = "AI Normal Mode";
+      message =
+        `Nice to meet you, ${userName}. You are playing AI normal mode.` +
+        "<br>" +
+        `Computer will help to play on your behalf.`;
+      return message;
+    } else if (input == 4) {
+      gameState = "playAI2";
+      currentMode = "AI Reverse Mode";
+      message =
+        `Nice to meet you, ${userName}. You are playing AI reverse mode.` +
+        "<br>" +
+        `Computer will help to play on your behalf.`;
+      return message;

```

Comment:
this is ok but you are not using the full benefits of early return. see how in line 24-28 you returned early and didnt have to use `else if` for lines 31-46. the same thing can be applied here!

recall that when we return, the rest of the function will stop running. the benefit of doing an early return is that we do not need to nest our logic in many layers of else if and else which can be unreadable and hard to follow esp if its a more complex project with many layers of nesting

Reply:
also seems like you're not using the intermediary `message` variable at all so you might not even need it. you can just return the string directly similar to how you're doing it in line 83